msgraphcore/middleware/retry_middleware.py

```diff
@@ -0,0 +1,206 @@
+import datetime
+import random
+import sys
+import time
+from email.utils import parsedate_to_datetime
+
+from msgraphcore.middleware.middleware import BaseMiddleware
+from msgraphcore.middleware.options.middleware_control import middleware_control
+
+
+class RetryMiddleware(BaseMiddleware):
+    """
+    TransportAdapter that allows us to specify the
+    retry policy for all requests
+    """
+
+    DEFAULT_TOTAL_RETRIES = 3
```

Comment:
We have since moved away from using `MiddlewareOptions` objects in our new design. Each middleware will fetch per request options from the `middleware_control` dictionary in the request context.